Assets/Stack mini game/Scripts/ChangeMovingCubeDirectionCollider.cs

```diff
@@ -0,0 +1,17 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class ChangeMovingCubeDirectionCollider : MonoBehaviour
+{
+    public Action collisionWithMovingCubeEvent;
+
+
+    private void OnTriggerEnter(Collider other)
+    {
+        //Debug.Log("Collision with " + other);
+        collisionWithMovingCubeEvent.Invoke();

```

Comment:
You know what. I have never actually looked into the performance aspect of the Events. I will definitely do that, this or next week. 

The reason I use them here is for decoupling purposes. I like that the instances of the ChangeMovingCubeDirectionCollider don't need to call the CubeMover's ChangeDirection. Because the way I see it, this way each class has its own responsibility: ChangeMovingCubeDirectionCollider just tell other classes if something triggered their collider, and on the other hand CubeMover moves the cube and changes its direction.

Reply:
The way I see it, what is happening here is exactly the same as just calling the method, it just happens via callback. But I dunno if that's correct. They still have separate responsibilities. They are separate modules. It's 

Because Unity doesn't use dependency injection, I feel like it's impossible to "decouple" them in the way you are trying to do.

Reply:
Performance is fine here of course, as its such small scope. but yeah worth a read. 

In Unity, maybe I am wrong. But I can't imagine why.

Reply:
Maybe you're right that it separates the component somewhat thinking about it lol. Now I'm not so sure if what I said is correct :D. I was thinking cos you don't need to actually create an instance of an object - like do ```new SomeObject()``` (as we are in Unity, not ASP.NET for instance, you just have a pre-existing monobehaviour in the scene that you point to). 

Actions are normally used when you have multiple things subscribing to them though, and you handle the invocation all in one place. Like for instance, my OnEndOfDay event in the CalendarManager. After each day, the event is fired, and then all the subscribers do stuff that relates to that scenario. Kinda like the observer pattern. 

Whereas here it's 1 neighbouring thing listening for another neighbouring thing's event? I should really read more about Actions, but normally from what I've seen they're used in a wider design pattern, e.g. Observer.